File: src/adaptive/adaptive_single_sample/adaptive_single_sample.py
import numpy as np

class AdaptiveFilterSingleSample:
    def __init__(self, num_taps: int, mu: float, num_channels: int = 3):
        """
        Initialize the multi-channel adaptive LMS filter.
        
        Args:
            num_taps (int): The number of filter taps (filter length)
            mu (float): The learning rate (step size)
            num_channels (int): Number of input channels (default 3 for accelerometer)
        """
        self.num_taps = num_taps
        self.mu = mu
        self.num_channels = num_channels
        
        # Initialize weights and buffers for each channel
        self.weights = np.zeros((num_channels, num_taps))
        self.buffer = np.zeros((num_channels, num_taps))

    def adapt(self, x: list, desired_signal: float):
        """
        Adapt the filter for each channel.
        
        Args:
            x (list): Current input samples [x, y, z] from accelerometer
            desired_signal (float): Current desired sample
        
        Returns:
            outputs (list): Filter outputs for each channel
            errors (list): Error signals for each channel
        """
        if len(x) != self.num_channels:
            raise ValueError(f"Expected {self.num_channels} channels, got {len(x)}")

        outputs = np.zeros(self.num_channels)
        errors = np.zeros(self.num_channels)

        for channel in range(self.num_channels):
            # Shift buffer and add new sample for current channel
            self.buffer[channel, 1:] = self.buffer[channel, :-1]
            self.buffer[channel, 0] = x[channel]

            # print("Channel:", channel)
            # print("Buffer:", self.buffer[channel])
            # print("Desired signal:", desired_signal)
            # Compute output for current channel
            outputs[channel] = np.dot(self.weights[channel], self.buffer[channel])

            # print("Output signal:", outputs[channel])
            # Calculate error
            errors[channel] = desired_signal - outputs[channel]
            
            # print("Error signal:", errors[channel])
            # Update weights for current channel
            self.weights[channel] += self.mu * errors[channel] * self.buffer[channel]

        #     print("Weights:", self.weights[channel])
        # print("-----------------")
        return outputs.tolist()

File: src/adaptive/adaptive_single_sample/c/adaptive_single_sample.c
#include "adaptive_single_sample.h"
#include <stdlib.h>
#include <stdio.h>

AdaptiveFilter *filter_create(int num_taps, float mu, int num_channels) {
    AdaptiveFilter *filter = malloc(sizeof(AdaptiveFilter));
    if (!filter)
        return NULL;

    filter->num_taps = num_taps;
    filter->mu = mu;
    filter->num_channels = num_channels;
    filter->iteration = 0;

    filter->weights = malloc(num_channels * sizeof(float *));
    filter->buffer = malloc(num_channels * sizeof(float *));

    for (int i = 0; i < num_channels; i++) {
        filter->weights[i] = calloc(num_taps, sizeof(float));
        filter->buffer[i] = calloc(num_taps, sizeof(float));
    }

    return filter;
}

void filter_adapt(AdaptiveFilter *filter, float *x, float desired_signal, float *outputs) {
    printf("Iteration %d\n", filter->iteration);
    filter->iteration++;

    for (int channel = 0; channel < filter->num_channels; channel++) {
        // Shift buffer
        for (int i = filter->num_taps - 1; i > 0; i--) {
            filter->buffer[channel][i] = filter->buffer[channel][i - 1];
        }
        filter->buffer[channel][0] = x[channel];
        
        // Print the buffer
        for (int i = 0; i < filter->num_taps; i++) {
            printf("  Buffer[%d][%d]: %.10f\n", channel, i, filter->buffer[channel][i]);
        }
        // print the desired signal
        printf("  Desired: %.10f\n", desired_signal);

        // Print the weights
        for (int i = 0; i < filter->num_taps; i++) {
            printf("  Weights[%d][%d]: %.10f\n", channel, i, filter->weights[channel][i]);
        }

        // Compute output
        outputs[channel] = 0;
        for (int i = 0; i < filter->num_taps; i++) {
            outputs[channel] += filter->weights[channel][i] * filter->buffer[channel][i];
        }

        // Print the output
        printf("  Output: %.10f\n", outputs[channel]);

        // Update weights
        float error = desired_signal - outputs[channel];

        // print the error
        printf("  Error: %.10f\n", error);

        for (int i = 0; i < filter->num_taps; i++) {
            filter->weights[channel][i] += filter->mu * error * filter->buffer[channel][i];
        }
        // print the updated weights
        for (int i = 0; i < filter->num_taps; i++) {
            printf("  Weights Updated[%d][%d]: %.10f\n", channel, i, filter->weights[channel][i]);
        }
        break;
    }
}

void filter_free(AdaptiveFilter *filter) {
    for (int i = 0; i < filter->num_channels; i++) {
        free(filter->weights[i]);
        free(filter->buffer[i]);
    }
    free(filter->weights);
    free(filter->buffer);
    free(filter);
}